### PR TITLE
Fixes to ctrl+backspace behavior

### PR DESCRIPTION
**Pull request by gremble0**
# Description
First contribution to chatterino, so hello :)

This PR addresses issue #5004 + fixes some issues related to that issue

## Changes
Added ctrl+backspace functionality while there is an active selection to `EmotePopup`, `SettingsDialog` and `SelectChannelDialog` classes, as well as a small change (fix) to ctrl+backspace behavior in and `SearchPopup`, which is also the behavior implemented in the other classes. 

Previously the behaviour for ctrl+backspace with an active selection was to clear all characters from the end of the selection to start of line (or the behavior described in the issue for the classes with it not implemented). This has been changed to ctrl+backspace deleting only the selected text, which is the behavior present in all other applications i tested it in.

Old behavior:
https://github.com/Chatterino/chatterino2/assets/45577341/160dd1e4-f46e-43aa-826e-a827dc264908

New behavior:
https://github.com/Chatterino/chatterino2/assets/45577341/8e2a4e4f-92e1-4f67-a54d-3d0544e1c7d9

It may be possible make a more generic solution to this issue than to add these `eventFilter()` methods wherever there is a input field, but as of yet i'm not familiar enough with QT and the chatterino codebase to do that.

### CODE REVIEW

File: src/widgets/dialogs/EmotePopup.cpp [hunk 1]
```diff
@@ -448,6 +449,22 @@ void EmotePopup::loadChannel(ChannelPtr channel)
...
 }
 
+bool EmotePopup::eventFilter(QObject *object, QEvent *event)
+{
+    if (object == this->search_ && event->type() == QEvent::KeyPress)
+    {
+        QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
+        if (keyEvent->key() == Qt::Key_Backspace &&
+            keyEvent->modifiers() == Qt::ControlModifier &&
+            this->search_->selectionLength() > 0)
```
Comment (Nerixyz) [inline src/widgets/dialogs/EmotePopup.cpp:459]:
```suggestion
        if (keyEvent == QKeySequence::DeleteStartOfWord &&
            this->search_->selectionLength() > 0)
```

We should be platform-agnostic.

File: src/widgets/dialogs/SelectChannelDialog.cpp [hunk 2]
```diff
@@ -497,6 +497,13 @@ bool SelectChannelDialog::EventFilter::eventFilter(QObject *watched,
             widget->previousInFocusChain()->setFocus();
             return true;
         }
+        else if (event_key->key() == Qt::Key_Backspace &&
+                 event_key->modifiers() == Qt::ControlModifier &&
+                 this->dialog->ui_.twitch.channelName->selectionLength() > 0)
```
Comment (Nerixyz) [inline src/widgets/dialogs/SelectChannelDialog.cpp:502]:
```suggestion
        else if (event_key == QKeySequence::DeleteStartOfWord &&
                 this->dialog->ui_.twitch.channelName->selectionLength() > 0)
```

Same here.

File: src/widgets/dialogs/SettingsDialog.cpp [hunk 3]
```diff
@@ -206,6 +207,22 @@ void SettingsDialog::setElementFilter(const QString &query)
...
 }
 
+bool SettingsDialog::eventFilter(QObject *object, QEvent *event)
+{
+    if (object == this->ui_.search && event->type() == QEvent::KeyPress)
+    {
+        QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
+        if (keyEvent->key() == Qt::Key_Backspace &&
+            keyEvent->modifiers() == Qt::ControlModifier &&
+            this->ui_.search->selectionLength() > 0)
```
Comment (Nerixyz) [inline src/widgets/dialogs/SettingsDialog.cpp:217]:
```suggestion
        if (keyEvent == QKeySequence::DeleteStartOfWord &&
            this->ui_.search->selectionLength() > 0)
```

And here.

File: src/widgets/helper/SearchPopup.cpp [hunk 4]
```diff
@@ -204,9 +204,9 @@ bool SearchPopup::eventFilter(QObject *object, QEvent *event)
         QKeyEvent *keyEvent = static_cast<QKeyEvent *>(event);
         if (keyEvent->key() == Qt::Key_Backspace &&
             keyEvent->modifiers() == Qt::ControlModifier &&
```
Comment (Nerixyz) [inline src/widgets/helper/SearchPopup.cpp:206]:
```suggestion
        if (keyEvent == QKeySequence::DeleteStartOfWord &&
```

And here.

**Review by Nerixyz — COMMENTED** [review]
This is probably a bug in Qt. I couldn't find a bugreport after a quick search.

The bug is in [`QWidgetLineControl::processKeyEvent`](https://github.com/qt/qtbase/blob/dev/src/widgets/widgets/qwidgetlinecontrol.cpp#L1816-L1821). This calls `cursorWordBackward`, which removes the selected word from the selection. Calling `del` will only delete the remaining selection.

**Comment by gremble0** [comment]
Yeah, i supposed this sounds like something chatterino probably shouldn't have to deal with. Anyways I've added your suggestions which will fix the behavior until it potentially gets fixed upstream. Until then i suppose it's up to you what you want to do in chatterino.

**Comment by pajlada** [comment]
@Nerixyz Does it make sense to merge this in the interim, while waiting for your fix to Qt to be released?
I'm guessing yes since this fix is small enough that it would be easy to revert once we're on a fixed Qt release

**Comment by Nerixyz** [comment]
> @Nerixyz Does it make sense to merge this in the interim, while waiting for your fix to Qt to be released? I'm guessing yes since this fix is small enough that it would be easy to revert once we're on a fixed Qt release

Yes, we should merge this. The fix should be in 6.6.2 (to be released on 17.01.2024). But, at least on Windows, we're still waiting on [509236](https://codereview.qt-project.org/c/qt/qtbase/+/509236).

**Review by pajlada — APPROVED** [review]

**Comment by pajlada** [comment]
Thank you! As a first-time contributor, you can now add yourself to the contributors list that's shown inside the About page in Chatterino.

If you want this, you can open a new PR where you modify the `resources/contributors.txt` file and add yourself to the list. Make sure to read the comments at the top for instructions.

**Pull request merged**